mono/metadata/mempool.c

```diff
@@ -334,22 +464,29 @@ mono_mempool_alloc0 (MonoMemPool *pool, guint size)
 {
 	gpointer rval;
 
+	guint size_with_canaries = size + canary_count * sizeof(guint8 *);
+
 #ifdef MALLOC_ALLOCATION
 	rval = mono_mempool_alloc (pool, size);
 #else
-	size = ALIGN_SIZE (size);
+	size_with_canaries = ALIGN_SIZE (size_with_canaries);
 
 	rval = pool->pos;
-	pool->pos = (guint8*)rval + size;
+	pool->pos = (guint8*)rval + size_with_canaries;
 
 	if (G_UNLIKELY (pool->pos >= pool->end)) {
+		// Don't double-pad with canaries
 		rval = mono_mempool_alloc (pool, size);
-	}
+	} else {
 #ifdef TRACE_ALLOCATIONS
-	else if (pool == mono_get_corlib ()->mempool) {
-		mono_backtrace (size);
-	}
+		if (pool == mono_get_corlib ()->mempool) {
```

Comment:
It's being indented because it's inside of a new else block. I can unindent the code inside the ifdef block though, if that's the preferred style here.